### ISSUE TITLE
PVC Storage Capacity Issue

### ISSUE THREAD

**Issue by warrenchristian1telus** [bcgov/digital-journeys#1721]
I've been alerted to a storage capacity issue in production (as well as dev and test).

The issue that we're using 10 GB of storage, but have allocated 63 GB to contain it.

<img width="757" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMTlWQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8b4bb7b060eb922e2628f5920462941e3d2fe1e3/image.png" alt="image.png" />

This is the "requested.storage" capacity limit, and as far as I can tell it won't actually impact anything unless we try to add another drive (PVC). This could currently happen with Keycloak being an Auto-scaler. It may auto-create PVC drives for each instance. Scaling a database StatefulSet will definitely do that.

Prod and dev are both 97.7% full, test is 82% - which is OK, but not quite sure why it's smaller.

I checked the new namespace as well (e66ac2-test) and it's at 82% (matching the main test deployment). Curious why there is a mismatch like that. May be a remnants of the initial deployments or manual adjustments. Or possibly extra PVC's were created in dev/prod due to service/pod scaling.

We'll need to discuss our options / solutions for this. I don't expect it needs to be a major rush, as it shouldn't crash anything. It could block any deployments or Auto-scalers that try to create a new disk drive. Say, if Keycloak tries to scale-up due to high traffic (as it is using Auto-scale).

I expect this will be a somewhat difficult/complex fix.

Unfortunately, reducing the size a PVC entails deleting it entirely, after removing any pod/service it is connected to. Then redeploying, recreating and restoring from backups.

Basically, a comprehensive test of our disaster recovery protocols in every namespace.

Could be a great opportunity to ensure the procedure and documentation are up to snuff. :)

We will definitely need to do something to resolve this before we can grow, scale, or add any redundancies (such as 3x each pod) for enhanced resilience.

I suppose an alternate option could be to request a storage increase, along with the CPU and memory increases we're already planning to request.

The issue with a storage increase is that the problem is exponential. If each drive is 5x larger than required (which they roughly are now), each time we scale up things will quickly get overloaded. To duplicate each pod, we need 5x the capacity it uses...and of course this is reflected through dev, test and prod. So each additional pod is equivalent to adding 15 in terms of required storage capacity compared to usage on the cluster.

Best practise is to right-size all the pods with what we've learned from their use thus far.

TL;DR:

At this point, I think a better rule of thumb would be to shoot for 2x the capacity in the majority of cases. This would save us 3x per pod, x3 namespaces, x3 pods per namespace; while also providing a buffer for growth and scaling.

PVC storage can easily be scaled-up if necessary, but it's a bit more difficult to scale-down. However, I did some calculations and I think we can save a significant amount of storage capacity by right-sizing the PVC's.

With 13 pods currently running in 3 namespaces at 5x capacity, once we 3x pods for redundancy (13 x 3 x 5 x 3 = 585), if we then subtract the capacity of the reduced-size (2x) pods, (13 x 3 x 2 x 3 = 234) would mean 585 - 234 = 351 pods worth of capacity would be saved saved in our next round of updates.

I expect we would reduce our current requested capacity from 63 GB (and 97% full) to under 20 GB per namespace. The total for all namespaces [dev, test, prod] is ~190 GB, which would be reduced to 60 GB for the entire project. Less than we use for each namespace today.

That would reduce projected capacity requirements for the project after our next upgrade from 750 GB down to to 180 GB, saving around 570 GB. It should also allow us to scale up to 3x pods in each namespace without requiring any additional requested.storage capacity.

I believe added efficiency and cost savings will be worth the effort to right-size the PVC's in the long-run, and will likely be required to scale the project at all.

**Comment by ayushdamani** [comment]
- Look at our backups in each env and figure out what we need to restore (old forms, tests etc)
- Making sure we have a good documentation for recovery - @bhumin-fw do we have such documentation?

**Comment by Stella-Archer** [comment]
Kanal will write action that removed unessary files form upload 
File storage reduction > Kanal
PVC in mongo/postgsress
telework attachments aren't deleted - 120 day - is file older delete

**Comment by Stella-Archer** [comment]
@warrenchristian1telus  this ticket need to be broken down into steps and stages

**Comment by Stella-Archer** [comment]
@warrenchristian1telus  will request more CPU so we don't risk crashing Cold and Flu launch June 10

**Comment by Stella-Archer** [comment]
@Stella-Archer will move this into an epic

**Comment by Stella-Archer** [comment]
Testing ticket # 
data form last year # forms/day average
what happened with the cold and flu crash last year
maximum busiest times

**Comment by Stella-Archer** [comment]
@warrenchristian1telus have you requested more CPU?

**Comment by warrenchristian1telus** [comment]
Not yet @Stella-Archer - more CPU in itself won't resolve our issue. We will need to properly allocate it, and possibly introduce additional scaling. I believe we'll need to finish mapping, optimizing and testing in dev to figure out how much CPU, memory, etc. we need to put where exactly.

For an more detailed explanation of what is required to make the increase request, please see: [Requesting a quota increase for an OpenShift project set](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/automation-and-resiliency/request-quota-increase-for-openshift-project-set/)

We'll also need to build and run effective load tests to measure improvements before deploying to production. It would be better if we could allow for load tests of production as well, because as much as we try, no two systems are ever quite the same under load (dev, test, prod) so performance adjustments can sometimes have varying effectiveness across deployments. This mainly requires the kitchen sink form being accessible to anyone. It will also require building a test or two to match.

I think we should probably add a few tasks to the list:

- [ ] Evaluate resource usage: Analyze your project's resource consumption to determine if the quota increase is necessary. If so, determine how much is required.
- [ ] Develop and deploy load tests to run against the Kitchen Sink form. Profile and track resource and performance data to identify quota and performance issues / solutions.
- [ ] [Application resource tuning](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/automation-and-resiliency/application-resource-tuning/)
- [ ] [Application resiliency](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/automation-and-resiliency/app-resiliency-guidelines/)

Within that cycle, we should be able to determine and prove the resources we require, as well as acquire the evidence to submit for review.

On a side-note, I think implementing Snowplow would go a long way to determining real usage and creating appropriate load tests. It should basically be a copy-paste to integrate. Not sure if there is another holding back implementation: https://github.com/bcgov/digital-journeys/issues/1352

**Comment by fazil-ey** [comment]
1. An issue with backups for files/pdfs - which is being sorted 
2. Infrastructure as code to build there environments automatically, give devs the ability to make adjustments - requires more time than doing it manually once

**Comment by fazil-ey** [comment]
1. An issue with backups for files/pdfs - which is being sorted 
2. Infrastructure as code to build there environments automatically, give devs the ability to make adjustments - requires more time than doing it manually once
